crates/ruff_server/src/server/schedule/thread/pool.rs

```diff
@@ -0,0 +1,106 @@
+// +------------------------------------------------------------+
+// | Code adopted from:                                         |
+// | Repository: https://github.com/rust-lang/rust-analyzer.git |
+// | File: `crates/stdx/src/thread/pool.rs`                     |
+// | Commit: 03b3cb6be9f21c082f4206b35c7fe7f291c94eaa           |
+// +------------------------------------------------------------+
+//! [`Pool`] implements a basic custom thread pool
+//! inspired by the [`threadpool` crate](http://docs.rs/threadpool).
+//! When you spawn a task you specify a thread priority
+//! so the pool can schedule it to run on a thread with that priority.
+//! rust-analyzer uses this to prioritize work based on latency requirements.
+//!
+//! The thread pool is implemented entirely using
+//! the threading utilities in [`crate::server::schedule::thread`].
+
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    Arc,
+};
+
+use crossbeam::channel::{Receiver, Sender};
+
+use super::{Builder, JoinHandle, ThreadPriority};
+
+pub(crate) struct Pool {
+    // `_handles` is never read: the field is present
+    // only for its `Drop` impl.
+
+    // The worker threads exit once the channel closes;
+    // make sure to keep `job_sender` above `handles`
+    // so that the channel is actually closed
+    // before we join the worker threads!
+    job_sender: Sender<Job>,
+    _handles: Vec<JoinHandle>,
+    extant_tasks: Arc<AtomicUsize>,
+}
+
+struct Job {
+    requested_priority: ThreadPriority,
+    f: Box<dyn FnOnce() + Send + 'static>,
+}
+
+impl Pool {
+    pub(crate) fn new(threads: usize) -> Pool {
+        const STACK_SIZE: usize = 8 * 1024 * 1024;
+        const INITIAL_PRIORITY: ThreadPriority = ThreadPriority::Worker;
+
+        let (job_sender, job_receiver) = crossbeam::channel::unbounded();
```

Comment:
Do we need unbounded channels here? I'd rather we use bounded channels so that we can benefit from backpressure. (And if we need unbounded channels, we should write some comments explaining why.)

Reply:
Yes! I wrote a little bit about it (with links) in the context of `uv` a bit ago: https://github.com/astral-sh/uv/pull/1163#discussion_r1473155361

To be clear, I do not mean to say "never use unbound channels." I mean to say that if we do use them, IMO, we ought to have a specific justification for them. But otherwise default to bounded channels.